pkg/logs/input/container/filter.go

```diff
@@ -0,0 +1,145 @@
+		return source
+	}
+	var candidate *config.LogSource
+	for _, source := range sources {
+		if source.Config.Image != "" && !isImageMatch(source.Config.Image, container.Image) {
+			continue
+		}
+		if source.Config.Label != "" && !isLabelMatch(source.Config.Label, container.Labels) {
+			continue
+		}
+		if candidate == nil {
+			candidate = source
+		}
+		if computeScore(container, candidate) < computeScore(container, source) {
+			candidate = source
+		}
+	}
+	return candidate
+}
+
+// computeScore returns the matching score between the container and the source,
+// the higher, the better
+func computeScore(container types.Container, source *config.LogSource) int {
+	score := 0
+	if isImageMatch(source.Config.Image, container.Image) {
+		score++
+	}
+	if isLabelMatch(source.Config.Label, container.Labels) {
+		score++
+	}
+	return score
+}
+
+// digestPrefix represents a prefix that can be added to an image name.
+const digestPrefix = "@sha256:"
+
+// isImageMatch returns true if image respects format: "Y/imageFilter@sha256:X" with 'Y/' and '@sha256:X' optional.
+func isImageMatch(imageFilter string, image string) bool {
+	if strings.Contains(image, digestPrefix) {
+		// Trim digest if present
+		splitted := strings.SplitN(image, digestPrefix, 2)
+		image = splitted[0]
+	}
+	// Expect prefix to end with '/'
+	repository := strings.TrimSuffix(image, imageFilter)
+	return len(repository) == 0 || strings.HasSuffix(repository, "/")
+}
+
+// isLabelMatch returns true if labels contains at least one label from labelFilter.
+func isLabelMatch(labelFilter string, labels map[string]string) bool {
+	// Expect a comma-separated list of labels, eg: foo:bar, baz
+	for _, value := range strings.Split(labelFilter, ",") {
+		// Trim whitespace, then check whether the label format is either key:value or key=value
+		label := strings.TrimSpace(value)
+		parts := strings.FieldsFunc(label, func(c rune) bool {
+			return c == ':' || c == '='
+		})
+		// If we have exactly two parts, check there is a container label that matches both.
+		// Otherwise fall back to checking the whole label exists as a key.
+		if _, exists := labels[label]; exists || len(parts) == 2 && labels[parts[0]] == parts[1] {
+			return true
+		}
+	}
+	return false
+}
+
+// logsConfigPath refers to the logs configuration that can be passed over a docker label,
+// this feature is commonly named autodicovery.
+const logsConfigPath = "com.datadoghq.ad.logs"
+
+// sourceFromContainer returns the source extracted and computed from the container labels.
+func sourceFromContainer(container types.Container) *config.LogSource {
+	logsConfig := extractLogsConfig(container.Labels)
+	if logsConfig == nil {
+		return nil
+	}
+	return config.NewLogSource(logsConfigPath, logsConfig)
+}
+
+// extractLogsConfig returns the logs config present in the label 'com.datadoghq.ad.logs',
+// the config has to be conform with the format '[{...}]'.
+func extractLogsConfig(labels map[string]string) *config.LogsConfig {
```

Comment:
I think you could improve the quality of your code by refining the encapsulation of your variables.
I'll take this one as an example but there are others around:
- `extractLogsConfig(labels map[string]string)` lacks cohesion because the caller leaks `.Labels`. How is a label map related to a config ? It should be an implementation detail not visible from the method signature.
- `extractLogsConfig(container types.Container)` is better (more self-contained).
The method name could still be improved.

By the way, I don't think we need to repeat `Logs` all over the place, we already are in a `logs` pkg.